### PR TITLE
Unittests Controller

### PR DESCRIPTION
**Pull request by Kevin-uxpkw**
Unittests so grob erstmal fertig, added export_raw_data in ProjectManager und save in AbstractController geändert und in ProjectManager ausgelagert um bei undo und redo zu nutzen

### CODE REVIEW

**Review by alice221927 — CHANGES_REQUESTED** [review]
Overall...Good
But i am not satisfied with the coverage (52%) :(

**Comment by alice221927** [comment]
Other issues with your code:
1. please make get thresholds return {} as initial and not sth. like {col1 :0, col2: 0}
2. You might have changed something in the evaluation controller because no no evaluation is possible!!

**Comment by alice221927** [comment]
This happens when i click on evaluate, even all the data is correct:

Traceback (most recent call last):
  File "C:\Users\Alissa\PycharmProjects\discrete-choice\src\model\ProxyProject.py", line 43, in __do_operation
    ret = func(np, *args, **kwargs)
  File "C:\Users\Alissa\PycharmProjects\discrete-choice\src\model\ProxyProject.py", line 113, in evaluate
    return self.evaluate()
  File "C:\Users\Alissa\PycharmProjects\discrete-choice\src\model\ProjectSnapshot.py", line 75, in evaluate
    self.__evaluation = self.__processing_configs[self.get_selected_config_index()].process(self.__model)
  File "C:\Users\Alissa\PycharmProjects\discrete-choice\src\model\processing\SingleLogitBiogemeConfig.py", line 30, in process
    model.data.complete_data.select_dtypes(exclude=['object']).dropna(axis=1, how='any'))
  File "C:\Users\Alissa\AppData\Local\Programs\Python\Python310\lib\functools.py", line 981, in __get__
    val = self.func(instance)
  File "C:\Users\Alissa\PycharmProjects\discrete-choice\src\model\data\Data.py", line 56, in complete_data
    row_value_dict = complete_data.to_dict(orient='index')
  File "C:\Users\Alissa\PycharmProjects\discrete-choice\venv\lib\site-packages\pandas\core\frame.py", line 1988, in to_dict
    return to_dict(self, orient, into, index)
  File "C:\Users\Alissa\PycharmProjects\discrete-choice\venv\lib\site-packages\pandas\core\methods\to_dict.py", line 177, in to_dict
    raise ValueError("DataFrame index must be unique for orient='index'.")
ValueError: DataFrame index must be unique for orient='index'.

**Comment by alice221927** [comment]
I got false coverage value. My bad.
The only changes, that you need to apply now are:
- Fix Undo Redo, so save get executed (The cause for the bug might also be in Model)
- Fix Evaluate (Take a look at your earlier versions of your branch and try to find when the problem started and what changes caused that)
- Fix Export (Currently suspecting the cause for the bug might also be in Model)

**Comment by einDamian** [comment]
I think it is better if separate issues are fixed outside of this test branch in their own branch.
So undo/redo and evaluate should be fixed here (if they work outside of this branch)
and the export should be a new branch (if you mean 'Export Data').

File: src/controller/ProjectManager.py [hunk 1]
```diff
@@ -365,4 +382,13 @@ def import_raw_data(self, path: str):
         :param path: path where the raw_data is located
         """
         raw_data = DataFrame(FileManager.import_(path))
+        raw_data = raw_data.set_index(raw_data.columns[0])
```
Comment (einDamian) [inline src/controller/ProjectManager.py:385]:
```suggestion
```
This line causes the first attribute to be interpreted as the line index.
This is false and causes problems when exporting.

File: src/controller/ProjectManager.py [hunk 2]
```diff
@@ -365,4 +382,13 @@ def import_raw_data(self, path: str):
...
         self.get_project().set_raw_data(raw_data, path)
+
+    def export_raw_data(self, path: str):
+        """
+        Exports the raw_data
+        :param path: path where the raw_data is exported to
+        """
+        raw_data = self.get_project().get_raw_data()
```
Comment (einDamian) [inline src/controller/ProjectManager.py:393]:
```suggestion
        raw_data = self.get_project().get_raw_data(with_derivatives=True)
```
Derivatives have to be added as this is currently only called when pressing `Export Data`.

**Review by einDamian — CHANGES_REQUESTED** [review]
With these changes `Export Data` should be functionable.

File: src/controller/ProjectManager.py (hunk 2 above)
Comment (alice221927) [reply src/controller/ProjectManager.py:393]: I tried this solution before (only changing get_raw_data() to get_raw_data(True))
See issue #80 for further details
It didn’t work

**Review by FliFlaFloo — APPROVED** [review]

**Review by einDamian — APPROVED** [review]

**Review by alice221927 — APPROVED** [review]

**Pull request merged**